### PR TITLE
Support vectorizing input slicing

### LINKED ISSUE #540: Vectorization of loads in horizontally fused `slice` operations
This is an issue for evaluation by Jie.

The expected SOL of the kernel on A100 is `(16*128*3072 * 2 * 4) / 1935*1e9 = 26 us`.  The measured kernel time is `~63us` giving a `41 %SOL`.  Notably, none of the loads are vectorized and only one of the stores is vectorized.  There is an odd transpose on one of the output that gives a transpose on the inner most dimension that could be complicating things.

```
import torch
from nvfuser import FusionDefinition, DataType

inputs = [
    torch.randn(16, 128, 3072, device='cuda'),
]

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.ops.slice(T0, start_indices=[0, 0, 0], end_indices=[16, 128, 1024], strides=[1, 1, 1])
    T2 = fd.ops.slice(T0, start_indices=[0, 0, 1024], end_indices=[16, 128, 2048], strides=[1, 1, 1])
    T3 = fd.ops.slice(T0, start_indices=[0, 0, 2048], end_indices=[16, 128, 3072], strides=[1, 1, 1])
    T4 = fd.ops.reshape(T2, original_shape=[16, 128, 1024], new_shape=[16, 128, 16, 64])
    T5 = fd.ops.permute(T4, dims=[0, 2, 1, 3])
    T6 = fd.ops.reshape(T1, original_shape=[16, 128, 1024], new_shape=[16, 128, 16, 64])
    T7 = fd.ops.permute(T6, dims=[0, 2, 1, 3])
    T8 = fd.ops.reshape(T3, original_shape=[16, 128, 1024], new_shape=[16, 128, 16, 64])
    T9 = fd.ops.permute(T8, dims=[0, 2, 1, 3])
    T10 = fd.ops.permute(T5, dims=[0, 1, 3, 2])
    fd.add_output(T7)
    fd.add_output(T9)
    fd.add_output(T10)

with FusionDefinition() as fd:
    nvfuser_fusion_id1(fd)

for _ in range(5):
    out = fd.execute(inputs)
```

Profile:

```
 1733917997          64288     307    16   256     1   128     1     1       30         0.000         0.000                                                     NVIDIA A100 80GB PCIe (0)    1     7  CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, (int)3, (int)3>, CudaCodeGen::Tensor<float, (int)4,…
```
Kernel:

```
__global__ void kernel1(Tensor<float, 3, 3> T0, Tensor<float, 4, 4> T7, Tensor<float, 4, 4> T9, Tensor<float, 4, 4> T10) {                                                                                                                                                       [115/1811]
  NVFUSER_DEFINE_MAGIC_ZERO;
  int i0;  
  i0 = (T0.size[2] * T0.size[1]) * (((nvfuser_index_t)blockIdx.x) - 0);                                                                      
  int i1;          
  i1 = 4 * ((nvfuser_index_t)threadIdx.x);
  int i2;        
  i2 = (4 * ((nvfuser_index_t)blockDim.x)) * ((nvfuser_index_t)blockIdx.y);
  int i3;                                               
  i3 = i1 + i2;     
  int i4;                     
  i4 = 131072 * ((nvfuser_index_t)blockIdx.x);            
  int i5;                     
  i5 = (i1 + i4) + i2;     
  bool b6;        
  b6 = ((i1 + 3) + i2) < 131072;                  
  int i7;                                         
  i7 = (-131072 + i1) + i2;           
  bool b8;                     
  b8 = ((3 + i1) + i2) < 131072;
  if (b6) {          
    #pragma unroll
    for(nvfuser_index_t i9 = 0; i9 < 4; ++i9) {
      int i10;    
      i10 = i3 + (i9 + nvfuser_zero);                                                                                                        
      int i11;
      i11 = i10 % 8192;                                                                                                                      
      int i12;     
      i12 = i10 / 8192;
      int i13;   
      i13 = i11 % 64;
      int i14;
      i14 = i11 / 64;         
      float T2[1];  
      T2[0] = 0;                                            
      T2[0]
         = T0[((i0 + (T0.size[2] * (i14 - 0))) + (((64 * i12) + i13) - -1024))];
      float T13[1];
      T13[0]      
         = T2[0];                                 
      float T5[1];
      T5[0]                    
         = T13[0];
      float T17[1];
      T17[0]
         = T5[0];
      T10[(((i4 + (8192 * i12)) + (128 * i13)) + i14)]
         = T17[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  } else {
    #pragma unroll
    for(nvfuser_index_t i9 = 0; i9 < 4; ++i9) {
      int i15;
      i15 = i9 + nvfuser_zero;
      int i16;
      i16 = i3 + i15;
      int i17;
      i17 = i16 % 8192;
      int i18;
      i18 = i16 / 8192;
      int i19;
      i19 = i17 % 64;
      int i20;
      i20 = i17 / 64;
      bool b21;
      b21 = i7 < (-i15);
      float T2[1];
      T2[0] = 0;
      if (b21) {
        T2[0]
           = T0[((i0 + (T0.size[2] * (i20 - 0))) + (((64 * i18) + i19) - -1024))];
      }
      float T13[1];
      T13[0]
         = T2[0];
      float T5[1];
      T5[0]
         = T13[0];
      float T17[1];
      T17[0]
         = T5[0];
      if (b21) {
        T10[(((i4 + (8192 * i18)) + (128 * i19)) + i20)]
           = T17[0];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  if (b6) {
    #pragma unroll
    for(nvfuser_index_t i22 = 0; i22 < 4; ++i22) {
      int i23;
      i23 = i22 + nvfuser_zero;
      int i24;
      i24 = i3 + i23;
      int i25;
      i25 = i24 % 8192;
      float T3[1];
      T3[0] = 0;
      T3[0]
         = T0[((i0 + (T0.size[2] * ((i25 / 64) - 0))) + (((64 * (i24 / 8192)) + (i25 % 64)) - -2048))];
      float T12[1];
      T12[0]
         = T3[0];
      float T16[1];
      T16[0]
         = T12[0];
      T9[(i5 + i23)]
         = T16[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  } else {
    #pragma unroll
    for(nvfuser_index_t i22 = 0; i22 < 4; ++i22) {
      int i26;
      i26 = i22 + nvfuser_zero;
      int i27;
      i27 = i3 + i26;
      int i28;
      i28 = i27 % 8192;
      bool b29;
      b29 = i7 < (-i26);
      float T3[1];
      T3[0] = 0;
      if (b29) {
        T3[0]
           = T0[((i0 + (T0.size[2] * ((i28 / 64) - 0))) + (((64 * (i27 / 8192)) + (i28 % 64)) - -2048))];
      }
      float T12[1];
      T12[0]
         = T3[0];
      float T16[1];
      T16[0]
         = T12[0];
      if (b29) {
        T9[(i5 + i26)]
           = T16[0];
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
  }
  if (b6) {
    Array<float, 4, 4> T15;
    #pragma unroll
    for(nvfuser_index_t i30 = 0; i30 < 4; ++i30) {
      int i31;
      i31 = i3 + (i30 + nvfuser_zero);
      int i32;
      i32 = i31 % 8192;
      float T1[1];
      T1[0] = 0;
      T1[0]
         = T0[((i0 + (T0.size[2] * ((i32 / 64) - 0))) + (((64 * (i31 / 8192)) + (i32 % 64)) - 0))];
      float T11[1];
      T11[0]
         = T1[0];
      T15[i30]
         = T11[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
    loadLocalToGlobal<float, 4, false>( &T7[i5], &T15[0]);
  } else {
    Array<float, 4, 4> T15;
    #pragma unroll
    for(nvfuser_index_t i30 = 0; i30 < 4; ++i30) {
      int i33;
      i33 = i3 + (i30 + nvfuser_zero);
      int i34;
      i34 = i33 % 8192;
      float T1[1];
      T1[0] = 0;
      if (b8) {
        T1[0]
           = T0[((i0 + (T0.size[2] * ((i34 / 64) - 0))) + (((64 * (i33 / 8192)) + (i34 % 64)) - 0))];
      }
      float T11[1];
      T11[0]
         = T1[0];
      T15[i30]
         = T11[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO;
    if (b8) {
      loadLocalToGlobal<float, 4, false>( &T7[i5], &T15[0]);
    }
  }
}
```

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #540 

Extended the vectorization analysis and validation to support slice ops with fusion inputs. Currently, any ops with `resize` are treated as non-vectorizable. This PR relaxes it and enables vectorization of slices with fusion inputs. Note that the slice op is represented using `resize`, so this PR adds a new analysis just for patterns like:

```
auto tv0 = makeContigConcreteTensor({1024});
auto tv1 = slice(tv0, {{4, -4}});
```

Here, the slice op would be vectorized by a factor of 4.

There are several things that need to be done differently than the normal vectorization case:

1. Usually we insert a `set` op between a fusion input and its uses and use that `set` op to vectorize the load. In the case of slice, we don't want to read the whole domain but just the sliced range, so we shouldn't insert a set op but rather directly vectorize the slice op. If an input is sliced as well as read as usual, we still do `cacheAfter` but just for the normal read, which means `cacheAfter` needs some extension.
2. When a domain is sliced, not only the slice range must be divisible by a vectorization factor, but its start offset address must be aligned, which means the alignment validation needs to be extended.
3. The start offset of a slice op may affect its vectorization factor. If it's just sliced once, the start offset doesn't matter as long as its address is aligned. However, if it's sliced once but also read entirely as usual, the transformation of the input must allow vectorization with both cases, which means the start offset is also divisible by a vectorization factor. Similarly, if an input is sliced twice, a vectorization factor must be a divisible factor of the start offsets of both slice ops, i.e., greatest common divisor of the two slice offsets.
4. Vectorization analysis also needs to consider that when a domain is sliced, its outer adjacent domain is no longer contiguous.

So far, I haven't thought much about any other `resize` use cases, so although some of the naming indicate relevance with `resize`, they are at this point only for slices. Furthermore, slices with non fusion inputs are not supported, which means that the projection analysis would fail to identify projectable extents, so vectorization would be effectively disabled.

### CODE REVIEW

**Comment by naoyam** [comment]
!build

File: csrc/device_lower/lower2device.h [hunk 1]
```diff
@@ -167,6 +167,10 @@ class TORCH_CUDA_CU_API GpuLower : public NonCopyable {
     return vectorized_set_info_;
   }
 
+  auto& tensorOffsets() {
```
Comment (zasdfgbnm) [inline csrc/device_lower/lower2device.h:170]: Is this used?

**Comment by zasdfgbnm** [comment]
Is it possible to split this PR into:
1. Lowering changes only, tested with manual scheduling.
2. Vectorization analysis changes, stacked on top of 1, tested by automatic schedulers?

**Comment by naoyam** [comment]
> Is it possible to split this PR into:
> 
> 1. Lowering changes only, tested with manual scheduling.
> 2. Scheduler changes, stacked on top of 1, tested by automatic schedulers?

That's a good idea. Let me do that.

**Comment by naoyam** [comment]
> > Is it possible to split this PR into:
> > 
> > 1. Lowering changes only, tested with manual scheduling.
> > 2. Scheduler changes, stacked on top of 1, tested by automatic schedulers?
> 
> That's a good idea. Let me do that.

1 is done in #733. Will update this PR based on the feedback to #733. Making this draft for now.

**Comment by naoyam** [comment]
Deprecated
